cpp/src/arrow/csv/converter.cc

```diff
@@ -467,5 +536,43 @@ Status Converter::Make(const std::shared_ptr<DataType>& type,
   return Make(type, options, default_memory_pool(), out);
 }
 
+Result<std::shared_ptr<DictionaryConverter>> DictionaryConverter::Make(
+    const std::shared_ptr<DataType>& type, const ConvertOptions& options,
+    MemoryPool* pool) {
+  DictionaryConverter* ptr;
```

Comment:
Can't you explicit `std::shared_ptr<DictionaryConverter>` here and use `ptr = std::make_shared<CONVERTER_TYPE>` after?

Reply:
There are problems with macro expansion and `<>` characters which force the use of parentheses, but prevents using the resulting expansion as a template parameter (i.e. `std::make_shared<(Foo<Bar>)>` doesn't seem recognized syntax).